allennlp/data/dataset_readers/nlvr2_reader.py

```diff
@@ -0,0 +1,230 @@
+                        examples_dict[key]["identifier"],
+                        examples_dict[key]["image_ids"],
+                        None,
+                        None,
+                        only_predictions=True,
+                    )
+                )
+        return instances
+
+    @overrides
+    def _read(self, split: str):
+        text_file_path = os.path.join(self.text_path_prefix, split + ".json")
+        visual_file_path = os.path.join(self.visual_path_prefix, split + ".tsv")
+        visual_features = load_obj_tsv(visual_file_path, self.topk_images)
+        for img in visual_features:
+            self.image_data[img["img_id"]] = img
+        with open(text_file_path) as f:
+            examples = json.load(f)
+            for example in examples:
+                if (
+                    example["img0"] not in self.image_data
+                    or example["img1"] not in self.image_data
+                ):
+                    continue
+                yield self.text_to_instance(
+                    example["sent"],
+                    example["identifier"],
+                    [example["img0"], example["img1"]],
+                    example["label"],
+                )
+
+    @overrides
+    def text_to_instance(
+        self,  # type: ignore
+        question: str,
+        identifier: str,
+        image_ids: List[str],
+        denotation: str = None,
+        only_predictions: bool = False,
+    ) -> Instance:
+        if self.mask_prepositions_verbs:
+            doc = self.nlp(question)
+            prep_verb_starts = [
+                (token.idx, len(token))
+                for token in doc
+                if token.dep_ == "prep" or token.pos_ == "VERB"
+            ]
+            new_question = ""
+            prev_end = 0
+            for (idx, length) in prep_verb_starts:
+                new_question += question[prev_end:idx] + self._tokenizer.mask_token
+                prev_end = idx + length
+            new_question += question[prev_end:]
+            question = new_question
+        elif self.drop_prepositions_verbs:
+            doc = self.nlp(question)
+            prep_verb_starts = [
+                (token.idx, len(token))
+                for token in doc
+                if token.dep_ == "prep" or token.pos_ == "VERB"
+            ]
+            new_question = ""
+            prev_end = 0
+            for (idx, length) in prep_verb_starts:
+                new_question += question[prev_end:idx]
+                prev_end = idx + length
+            new_question += question[prev_end:]
+            question = new_question
+        tokenized_sentence = self._tokenizer.tokenize(question)
+        tokenized_sentence = [str(tok) for tok in tokenized_sentence]
+        sentence_field = TextField(
+            [Token(self._tokenizer.cls_token)]
+            + [Token(tok) for tok in tokenized_sentence]
+            + [Token(self._tokenizer.sep_token)],
```

Comment:
You should call `self._tokenizer.add_special_tokens()` for this.

Reply:
For the dataset_readers, do we want to create the reader for each dataset?  (VQA, GQA, refer expression), tasks, or models?

Reply:
Yes, each dataset would need a dataset reader.  If suitable readers already exist in other places, we can re-use as much code as possible (though we'd definitely have to write at least a little bit of code to get them into allennlp's data pipeline).